# <center> **Weather Report Forecasting Analysis** </center>
## <center> **Module 1: Pre-processing, Analysing Data using Python** </center>


***
## Introduction :

Weather prediction helps to determine future climate changes. It has a great impact on various aspects of human life. Weather forecasting is done by collecting data on the current state of the atmosphere and applying a scientific understanding of atmospheric processes to predict and analyse atmospheric progression. In the project, we have been provided with day wise weather data from Jan 2022 to July 2033 which is a predicted
data. 

![download.gif](https://cdn.dribbble.com/users/1046127/screenshots/3987643/media/f8536f8486d6a831bc8ceff562e6e6e6.gif)


### Objectives of Module 1 :

• Handling null values, deletion or transformation of irrelevant values.

• Datatype transformation, formatting, removing duplicates.

• To Correct the years for given data set

• Encoding data into suitable format i.e. UTF8

• To get a cleaned dataset csv file


### Variables present in the dataset :

Columns i.e. attributes present in the dataset are as follows :


• Date- Date for which predicted data is given

• Average temperature (°F) - Avg. temperature at the given day in °F

• Average humidity (%) - Avg. humidity at the given day in %

• Average dewpoint (°F) - Avg. dewpoint at the given day in °F

• Average barometer (in) - Avg. barometer reading at the given day. Barometers predict the weather by detecting changes in the air pressure

• Average windspeed (mph) - Avg. windspeed at the given day

• Average gust speed (mph) -  Avg. gust speed at the given date It is a sudden, brief increase in speed of the wind, usually less than 20 seconds

• Average direction (°degree) - Avg. direction of the wind in degree

• Rainfall for month (in) - Rainfall predicted for  the month

• Rainfall for year (in) - Rainfall predicted for the year

• Maximum rain per minute - Max. rain per minute

• Maximum temperature (°F) - Max. temperature predicted for the given day

• Minimum temperature (°F) -Min. temperature predicted for the given day

• Maximum humidity (%) - Max. humidity predicted for the given day

• Minimum humidity (%) - Min. humidity predicted for the given day

• Maximum pressure - Max. pressure predicted for the given day

• Minimum pressure - Min. pressure predicted for the given day

• Maximum wind speed (mph) - Max. wind speed predicted in mph

• Maximum gust speed (mph) - Max. gust speed predicted in mph

• Maximum heat index (°F) - Max. heat index

• Diff pressure - Diff presupressure predicted


### Importing necessary libraries :

We will import the required libraries such as pandas, numpy. Pandas will help in data analysis and manipulation, while numpy works with the arrays use for scientific computing. It provides a high-performance multidimensional array object, and tools.

In [1]:
import numpy as np                              #importing numpy as np
import pandas as pd                             #importing pandas as pd

### To read csv dataset :

We will import our data into the jupyter noterbook and use pandas to read csv file and to store it in a dataframe in order to clean and analyse as follows:

In [2]:
df = pd.read_excel ('weather_dataset_stage1.xls', skiprows=28)   #to read the csv file
pd.pandas.set_option('display.max_columns', None)                #to print the all columns
df.head(5)

Date  Temperature  Average humidity (%  Average dewpoint (°F  \
0   2022-01-01       '37.8'               '35.0'                '12.7'   
1  2022-01-02'       '43.2'               '32.0'                '14.7'   
2  2022-01-03'       '25.7'               '60.0'                '12.7'   
3  2022-01-04'        '9.3'               '67.0'                 '0.1'   
4  2022-01-05'       '23.5'               '30.0'                '-5.3'   

   Average barometer (in  Average windspeed (mph  Average gustspeed (mph  \
0                 '29.7'                  '26.4'                  '36.8'   
1                 '29.5'                  '12.8'                  '18.0'   
2                 '29.7'                   '8.3'                  '12.2'   
3                 '30.4'                   '2.9'                   '4.5'   
4                 '29.9'                  '16.7'                  '23.1'   

   Average direction (°deg  Rainfall for month (in Rainfall for year (in  \
0                  '274.0'                   '0.0'                 '0.0'   
1                  '240.0'                   '0.0'                 '0.0'   
2                  '290.0'                   '0.0'                 '0.0'   
3                   '47.0'                   '0.0'                 '0.0'   
4                  '265.0'                   '0.0'                 '0.0'   

   Maximum rain per minute  Maximum temperature (°F  Minimum temperature (°F  \
0                    '0.0'                   '40.0'                   '34.0'   
1                    '0.0'                   '52.0'                   '37.0'   
2                    '0.0'                   '41.0'                    '6.0'   
3                    '0.0'                   '19.0'                   '-0.0'   
4                    '0.0'                   '30.0'                   '15.0'   

   Maximum humidity (%  Minimum humidity (%  Maximum pressure  \
0                '4.0'               '27.0'          '29.762'   
1                '4.0'               '16.0'          '29.669'   
2                '8.0'               '35.0'          '30.232'   
3                '7.0'               '35.0'          '30.566'   
4                '5.0'               '13.0'          '30.233'   

   Minimum pressure  Maximum windspeed (mph  Maximum gust speed (mph  \
0          '29.596'                  '41.4'                   '59.0'   
1          '29.268'                  '35.7'                   '51.0'   
2           '29.26'                  '25.3'                   '38.0'   
3          '30.227'                  '12.7'                   '20.0'   
4          '29.568'                  '38.0'                   '53.0'   

   Maximum heat index (°F         Date1  Month          diff_pressure  
0                  '40.0'  '2022-01-01'   '01'  '0.16600000000000037'  
1                  '52.0'  '2022-01-02'   '01'   '0.4009999999999998'  
2                  '41.0'  '2022-01-03'   '01'   '0.9719999999999978'  
3                  '32.0'  '2022-01-04'   '01'  '0.33899999999999864'  
4                  '32.0'  '2022-01-05'   '01'   '0.6649999999999991'

### Basic Exploration of the dataset :
We will perform basic exploration of the dataset to understand data given.

In [3]:
#To know the size of database using shape function
df.shape

(3903, 23)

In [4]:
# Exploring first 10 rows of the dataset using head function
df.head(5) 

Date  Temperature  Average humidity (%  Average dewpoint (°F  \
0   2022-01-01       '37.8'               '35.0'                '12.7'   
1  2022-01-02'       '43.2'               '32.0'                '14.7'   
2  2022-01-03'       '25.7'               '60.0'                '12.7'   
3  2022-01-04'        '9.3'               '67.0'                 '0.1'   
4  2022-01-05'       '23.5'               '30.0'                '-5.3'   

   Average barometer (in  Average windspeed (mph  Average gustspeed (mph  \
0                 '29.7'                  '26.4'                  '36.8'   
1                 '29.5'                  '12.8'                  '18.0'   
2                 '29.7'                   '8.3'                  '12.2'   
3                 '30.4'                   '2.9'                   '4.5'   
4                 '29.9'                  '16.7'                  '23.1'   

   Average direction (°deg  Rainfall for month (in Rainfall for year (in  \
0                  '274.0'                   '0.0'                 '0.0'   
1                  '240.0'                   '0.0'                 '0.0'   
2                  '290.0'                   '0.0'                 '0.0'   
3                   '47.0'                   '0.0'                 '0.0'   
4                  '265.0'                   '0.0'                 '0.0'   

   Maximum rain per minute  Maximum temperature (°F  Minimum temperature (°F  \
0                    '0.0'                   '40.0'                   '34.0'   
1                    '0.0'                   '52.0'                   '37.0'   
2                    '0.0'                   '41.0'                    '6.0'   
3                    '0.0'                   '19.0'                   '-0.0'   
4                    '0.0'                   '30.0'                   '15.0'   

   Maximum humidity (%  Minimum humidity (%  Maximum pressure  \
0                '4.0'               '27.0'          '29.762'   
1                '4.0'               '16.0'          '29.669'   
2                '8.0'               '35.0'          '30.232'   
3                '7.0'               '35.0'          '30.566'   
4                '5.0'               '13.0'          '30.233'   

   Minimum pressure  Maximum windspeed (mph  Maximum gust speed (mph  \
0          '29.596'                  '41.4'                   '59.0'   
1          '29.268'                  '35.7'                   '51.0'   
2           '29.26'                  '25.3'                   '38.0'   
3          '30.227'                  '12.7'                   '20.0'   
4          '29.568'                  '38.0'                   '53.0'   

   Maximum heat index (°F         Date1  Month          diff_pressure  
0                  '40.0'  '2022-01-01'   '01'  '0.16600000000000037'  
1                  '52.0'  '2022-01-02'   '01'   '0.4009999999999998'  
2                  '41.0'  '2022-01-03'   '01'   '0.9719999999999978'  
3                  '32.0'  '2022-01-04'   '01'  '0.33899999999999864'  
4                  '32.0'  '2022-01-05'   '01'   '0.6649999999999991'

In [5]:
# Exploring last 10 rows of dataset using tail function.
df.tail(5) 

Date  Temperature  Average humidity (%  Average dewpoint (°F  \
3898  2022-07-25'       '62.8'               '60.0'                '48.1'   
3899  2022-07-26'       '60.6'               '68.0'                '48.9'   
3900  2022-07-27'       '61.7'               '64.0'                '47.4'   
3901  2022-07-28'       '60.5'               '61.0'                '45.3'   
3902  2022-07-28'       '60.5'               '61.0'                '45.3'   

      Average barometer (in  Average windspeed (mph  Average gustspeed (mph  \
3898                 '29.7'                   '2.5'                   '4.0'   
3899                 '29.8'                   '1.7'                   '2.9'   
3900                 '29.9'                   '2.2'                   '4.0'   
3901                 '29.7'                   '4.0'                   '6.2'   
3902                 '29.7'                   '4.0'                   '6.2'   

      Average direction (°deg  Rainfall for month (in Rainfall for year (in  \
3898                  '242.0'                  '0.33'                '4.47'   
3899                  '357.0'                  '0.33'                '4.47'   
3900                   '66.0'                  '0.33'                '4.47'   
3901                  '248.0'                  '0.35'                '4.49'   
3902                  '248.0'                  '0.35'                '4.49'   

      Maximum rain per minute  Maximum temperature (°F  \
3898                    '0.0'                   '69.2'   
3899                    '0.0'                   '71.9'   
3900                    '0.0'                   '77.3'   
3901                    '0.0'                   '75.6'   
3902                    '0.0'                   '75.6'   

      Minimum temperature (°F  Maximum humidity (%  Minimum humidity (%  \
3898                   '55.1'               '90.0'               '36.0'   
3899                   '50.5'               '90.0'               '40.0'   
3900                   '43.6'               '96.0'               '35.0'   
3901                   '46.0'               '94.0'               '35.0'   
3902                   '46.0'               '94.0'               '35.0'   

      Maximum pressure  Minimum pressure  Maximum windspeed (mph  \
3898          '29.781'          '29.645'                   '8.1'   
3899           '29.93'          '29.745'                  '11.5'   
3900          '29.941'          '29.781'                  '13.8'   
3901          '29.792'          '29.675'                  '17.3'   
3902          '29.792'          '29.675'                  '17.3'   

      Maximum gust speed (mph  Maximum heat index (°F         Date1  Month  \
3898                   '17.3'                  '77.5'  '2022-07-25'   '07'   
3899                   '15.0'                  '77.5'  '2022-07-26'   '07'   
3900                   '18.4'                  '78.2'  '2022-07-27'   '07'   
3901                   '26.5'                  '77.6'  '2022-07-28'   '07'   
3902                   '26.5'                  '77.6'  '2022-07-28'   '07'   

               diff_pressure  
3898   '0.13599999999999923'  
3899   '0.18499999999999872'  
3900   '0.16000000000000014'  
3901  '0.11700000000000088';  
3902  '0.11700000000000088';

In [6]:
#To check for basic information of the dataset such as null elements and datatypes of column, memory usage by data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3903 entries, 0 to 3902
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Date                      3903 non-null   object
 1    Temperature              3903 non-null   object
 2    Average humidity (%      3903 non-null   object
 3    Average dewpoint (°F     3903 non-null   object
 4    Average barometer (in    3903 non-null   object
 5    Average windspeed (mph   3903 non-null   object
 6    Average gustspeed (mph   3903 non-null   object
 7    Average direction (°deg  3903 non-null   object
 8    Rainfall for month (in   3903 non-null   object
 9   Rainfall for year (in     3903 non-null   object
 10   Maximum rain per minute  3903 non-null   object
 11   Maximum temperature (°F  3903 non-null   object
 12   Minimum temperature (°F  3903 non-null   object
 13   Maximum humidity (%      3903 non-null   object
 14   Minimum humidity (%    

In [7]:
#to check which column have missing values using isna()
df.isna().sum()

Date                        0
 Temperature                0
 Average humidity (%        0
 Average dewpoint (°F       0
 Average barometer (in      0
 Average windspeed (mph     0
 Average gustspeed (mph     0
 Average direction (°deg    0
 Rainfall for month (in     0
Rainfall for year (in       0
 Maximum rain per minute    0
 Maximum temperature (°F    0
 Minimum temperature (°F    0
 Maximum humidity (%        0
 Minimum humidity (%        0
 Maximum pressure           0
 Minimum pressure           0
 Maximum windspeed (mph     0
 Maximum gust speed (mph    0
 Maximum heat index (°F     0
Date1                       0
 Month                      0
 diff_pressure              0
dtype: int64

Using info(),isna() function we get to know following information about the given dataset:

• There are total 3902 entries i.e. there are 8618 rows and 23 columns.

• All column's datatypes are reading as object type which is incorrect.

• There are no null values present in dataset.

## Pre-processing and Data Cleaning using Python :

As a Data Analyst, data cleaning is inevitable process. It is the process of handling or removing incorrect, corrupted, incorrectly formatted, duplicate, or incomplete data within a dataset. It is a part of the data preparation process. Data cleansing allows for accurate, defensible data that generates reliable visualizations, models, and business decisions.

In our given raw dataset, there are some duplicates, trailed whitespaces, missing data and incorrect data. We will further process data cleaning part.

In [8]:
#to strip unnecessary char from both ends
df = df.apply(lambda x:x.str.strip("';'"))                      

In [9]:
df.tail()

Date  Temperature  Average humidity (%  Average dewpoint (°F  \
3898  2022-07-25         62.8                 60.0                  48.1   
3899  2022-07-26         60.6                 68.0                  48.9   
3900  2022-07-27         61.7                 64.0                  47.4   
3901  2022-07-28         60.5                 61.0                  45.3   
3902  2022-07-28         60.5                 61.0                  45.3   

      Average barometer (in  Average windspeed (mph  Average gustspeed (mph  \
3898                   29.7                     2.5                     4.0   
3899                   29.8                     1.7                     2.9   
3900                   29.9                     2.2                     4.0   
3901                   29.7                     4.0                     6.2   
3902                   29.7                     4.0                     6.2   

      Average direction (°deg  Rainfall for month (in Rainfall for year (in  \
3898                    242.0                    0.33                  4.47   
3899                    357.0                    0.33                  4.47   
3900                     66.0                    0.33                  4.47   
3901                    248.0                    0.35                  4.49   
3902                    248.0                    0.35                  4.49   

      Maximum rain per minute  Maximum temperature (°F  \
3898                      0.0                     69.2   
3899                      0.0                     71.9   
3900                      0.0                     77.3   
3901                      0.0                     75.6   
3902                      0.0                     75.6   

      Minimum temperature (°F  Maximum humidity (%  Minimum humidity (%  \
3898                     55.1                 90.0                 36.0   
3899                     50.5                 90.0                 40.0   
3900                     43.6                 96.0                 35.0   
3901                     46.0                 94.0                 35.0   
3902                     46.0                 94.0                 35.0   

      Maximum pressure  Minimum pressure  Maximum windspeed (mph  \
3898            29.781            29.645                     8.1   
3899             29.93            29.745                    11.5   
3900            29.941            29.781                    13.8   
3901            29.792            29.675                    17.3   
3902            29.792            29.675                    17.3   

      Maximum gust speed (mph  Maximum heat index (°F       Date1  Month  \
3898                     17.3                    77.5  2022-07-25     07   
3899                     15.0                    77.5  2022-07-26     07   
3900                     18.4                    78.2  2022-07-27     07   
3901                     26.5                    77.6  2022-07-28     07   
3902                     26.5                    77.6  2022-07-28     07   

            diff_pressure  
3898  0.13599999999999923  
3899  0.18499999999999872  
3900  0.16000000000000014  
3901  0.11700000000000088  
3902  0.11700000000000088

In [10]:
df.head()

Date  Temperature  Average humidity (%  Average dewpoint (°F  \
0  2022-01-01         37.8                 35.0                  12.7   
1  2022-01-02         43.2                 32.0                  14.7   
2  2022-01-03         25.7                 60.0                  12.7   
3  2022-01-04          9.3                 67.0                   0.1   
4  2022-01-05         23.5                 30.0                  -5.3   

   Average barometer (in  Average windspeed (mph  Average gustspeed (mph  \
0                   29.7                    26.4                    36.8   
1                   29.5                    12.8                    18.0   
2                   29.7                     8.3                    12.2   
3                   30.4                     2.9                     4.5   
4                   29.9                    16.7                    23.1   

   Average direction (°deg  Rainfall for month (in Rainfall for year (in  \
0                    274.0                     0.0                   0.0   
1                    240.0                     0.0                   0.0   
2                    290.0                     0.0                   0.0   
3                     47.0                     0.0                   0.0   
4                    265.0                     0.0                   0.0   

   Maximum rain per minute  Maximum temperature (°F  Minimum temperature (°F  \
0                      0.0                     40.0                     34.0   
1                      0.0                     52.0                     37.0   
2                      0.0                     41.0                      6.0   
3                      0.0                     19.0                     -0.0   
4                      0.0                     30.0                     15.0   

   Maximum humidity (%  Minimum humidity (%  Maximum pressure  \
0                  4.0                 27.0            29.762   
1                  4.0                 16.0            29.669   
2                  8.0                 35.0            30.232   
3                  7.0                 35.0            30.566   
4                  5.0                 13.0            30.233   

   Minimum pressure  Maximum windspeed (mph  Maximum gust speed (mph  \
0            29.596                    41.4                     59.0   
1            29.268                    35.7                     51.0   
2             29.26                    25.3                     38.0   
3            30.227                    12.7                     20.0   
4            29.568                    38.0                     53.0   

   Maximum heat index (°F       Date1  Month        diff_pressure  
0                    40.0  2022-01-01     01  0.16600000000000037  
1                    52.0  2022-01-02     01   0.4009999999999998  
2                    41.0  2022-01-03     01   0.9719999999999978  
3                    32.0  2022-01-04     01  0.33899999999999864  
4                    32.0  2022-01-05     01   0.6649999999999991

In [11]:
#to strip whitespaces if present from both end
df = df.apply(lambda x:x.str.strip())            
df.head()

Date  Temperature  Average humidity (%  Average dewpoint (°F  \
0  2022-01-01         37.8                 35.0                  12.7   
1  2022-01-02         43.2                 32.0                  14.7   
2  2022-01-03         25.7                 60.0                  12.7   
3  2022-01-04          9.3                 67.0                   0.1   
4  2022-01-05         23.5                 30.0                  -5.3   

   Average barometer (in  Average windspeed (mph  Average gustspeed (mph  \
0                   29.7                    26.4                    36.8   
1                   29.5                    12.8                    18.0   
2                   29.7                     8.3                    12.2   
3                   30.4                     2.9                     4.5   
4                   29.9                    16.7                    23.1   

   Average direction (°deg  Rainfall for month (in Rainfall for year (in  \
0                    274.0                     0.0                   0.0   
1                    240.0                     0.0                   0.0   
2                    290.0                     0.0                   0.0   
3                     47.0                     0.0                   0.0   
4                    265.0                     0.0                   0.0   

   Maximum rain per minute  Maximum temperature (°F  Minimum temperature (°F  \
0                      0.0                     40.0                     34.0   
1                      0.0                     52.0                     37.0   
2                      0.0                     41.0                      6.0   
3                      0.0                     19.0                     -0.0   
4                      0.0                     30.0                     15.0   

   Maximum humidity (%  Minimum humidity (%  Maximum pressure  \
0                  4.0                 27.0            29.762   
1                  4.0                 16.0            29.669   
2                  8.0                 35.0            30.232   
3                  7.0                 35.0            30.566   
4                  5.0                 13.0            30.233   

   Minimum pressure  Maximum windspeed (mph  Maximum gust speed (mph  \
0            29.596                    41.4                     59.0   
1            29.268                    35.7                     51.0   
2             29.26                    25.3                     38.0   
3            30.227                    12.7                     20.0   
4            29.568                    38.0                     53.0   

   Maximum heat index (°F       Date1  Month        diff_pressure  
0                    40.0  2022-01-01     01  0.16600000000000037  
1                    52.0  2022-01-02     01   0.4009999999999998  
2                    41.0  2022-01-03     01   0.9719999999999978  
3                    32.0  2022-01-04     01  0.33899999999999864  
4                    32.0  2022-01-05     01   0.6649999999999991

### Exploring  and handling columns :

In [12]:
#to explore columns present
df.columns                      

Index(['Date', ' Temperature', ' Average humidity (%', ' Average dewpoint (°F',
       ' Average barometer (in', ' Average windspeed (mph',
       ' Average gustspeed (mph', ' Average direction (°deg',
       ' Rainfall for month (in', 'Rainfall for year (in',
       ' Maximum rain per minute', ' Maximum temperature (°F',
       ' Minimum temperature (°F', ' Maximum humidity (%',
       ' Minimum humidity (%', ' Maximum pressure', ' Minimum pressure',
       ' Maximum windspeed (mph', ' Maximum gust speed (mph',
       ' Maximum heat index (°F', 'Date1', ' Month', ' diff_pressure'],
      dtype='object')

We will rename column names using .rename() for better readiablity.

In [13]:
#to rename columns
df.rename(columns={
  'Date': 'Date',
  ' Temperature': 'Average_temperature',
  ' Average humidity (%': 'Average_humidity',
  ' Average dewpoint (°F': 'Average_dewpoint',
  ' Average barometer (in': 'Average_barometer',
  ' Average windspeed (mph': 'Average_windspeed',
  ' Average gustspeed (mph': 'Average_gust_speed',
  ' Average direction (°deg': 'Average_direction_degree',
  ' Rainfall for month (in': 'Rainfall_for_month',
  'Rainfall for year (in': 'Rainfall_for_year',
  ' Maximum rain per minute': 'Maximum_rain_per_minute',
  ' Maximum temperature (°F': 'Maximum_temperature',
  ' Minimum temperature (°F': 'Minimum_temperature',
  ' Maximum humidity (%': 'Maximum_humidity',
  ' Minimum humidity (%': 'Minimum_humidity',
  ' Maximum pressure': 'Maximum_pressure',
  ' Minimum pressure': 'Minimum_pressure',
  ' Maximum windspeed (mph': 'Maximum_windspeed',
  ' Maximum gust speed (mph': 'Maximum_gust_speed',
  ' Maximum heat index (°F': 'Maximum_heat_index',
  'Date1': 'Date1',
  ' Month': 'Month',
  ' diff_pressure': 'Diff_pressure'
}, inplace=True)

In [14]:
#to check new column names
df.columns            

Index(['Date', 'Average_temperature', 'Average_humidity', 'Average_dewpoint',
       'Average_barometer', 'Average_windspeed', 'Average_gust_speed',
       'Average_direction_degree', 'Rainfall_for_month', 'Rainfall_for_year',
       'Maximum_rain_per_minute', 'Maximum_temperature', 'Minimum_temperature',
       'Maximum_humidity', 'Minimum_humidity', 'Maximum_pressure',
       'Minimum_pressure', 'Maximum_windspeed', 'Maximum_gust_speed',
       'Maximum_heat_index', 'Date1', 'Month', 'Diff_pressure'],
      dtype='object')

In [15]:
df.head()           

Date Average_temperature Average_humidity Average_dewpoint  \
0  2022-01-01                37.8             35.0             12.7   
1  2022-01-02                43.2             32.0             14.7   
2  2022-01-03                25.7             60.0             12.7   
3  2022-01-04                 9.3             67.0              0.1   
4  2022-01-05                23.5             30.0             -5.3   

  Average_barometer Average_windspeed Average_gust_speed  \
0              29.7              26.4               36.8   
1              29.5              12.8               18.0   
2              29.7               8.3               12.2   
3              30.4               2.9                4.5   
4              29.9              16.7               23.1   

  Average_direction_degree Rainfall_for_month Rainfall_for_year  \
0                    274.0                0.0               0.0   
1                    240.0                0.0               0.0   
2                    290.0                0.0               0.0   
3                     47.0                0.0               0.0   
4                    265.0                0.0               0.0   

  Maximum_rain_per_minute Maximum_temperature Minimum_temperature  \
0                     0.0                40.0                34.0   
1                     0.0                52.0                37.0   
2                     0.0                41.0                 6.0   
3                     0.0                19.0                -0.0   
4                     0.0                30.0                15.0   

  Maximum_humidity Minimum_humidity Maximum_pressure Minimum_pressure  \
0              4.0             27.0           29.762           29.596   
1              4.0             16.0           29.669           29.268   
2              8.0             35.0           30.232            29.26   
3              7.0             35.0           30.566           30.227   
4              5.0             13.0           30.233           29.568   

  Maximum_windspeed Maximum_gust_speed Maximum_heat_index       Date1 Month  \
0              41.4               59.0               40.0  2022-01-01    01   
1              35.7               51.0               52.0  2022-01-02    01   
2              25.3               38.0               41.0  2022-01-03    01   
3              12.7               20.0               32.0  2022-01-04    01   
4              38.0               53.0               32.0  2022-01-05    01   

         Diff_pressure  
0  0.16600000000000037  
1   0.4009999999999998  
2   0.9719999999999978  
3  0.33899999999999864  
4   0.6649999999999991

### Fixing Data Types of Columns :

We will change the data types of Date column to Datetime and other columns as Float.

In [16]:
#changing Date column to datetime
df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m-%d', errors='coerce')
df['Date'].dtype

dtype('<M8[ns]')

In [17]:
df.info()          #to check column info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3903 entries, 0 to 3902
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Date                      3901 non-null   datetime64[ns]
 1   Average_temperature       3903 non-null   object        
 2   Average_humidity          3903 non-null   object        
 3   Average_dewpoint          3903 non-null   object        
 4   Average_barometer         3903 non-null   object        
 5   Average_windspeed         3903 non-null   object        
 6   Average_gust_speed        3903 non-null   object        
 7   Average_direction_degree  3903 non-null   object        
 8   Rainfall_for_month        3903 non-null   object        
 9   Rainfall_for_year         3903 non-null   object        
 10  Maximum_rain_per_minute   3903 non-null   object        
 11  Maximum_temperature       3903 non-null   object        
 12  Minimum_temperature 

We can see Date1 and Month are unnecessary columns we can drop them.

In [18]:
#to drop columns
df = df.drop(columns=['Date1', 'Month'])

Now, we will change the data types of remaining columns excluding Date column.

In [19]:
exclude_col = ['Date']

In [20]:
# to  get a list of all column names, except for the excluded column
cols = [col for col in df.columns if col not in exclude_col]

In [21]:
cols

['Average_temperature',
 'Average_humidity',
 'Average_dewpoint',
 'Average_barometer',
 'Average_windspeed',
 'Average_gust_speed',
 'Average_direction_degree',
 'Rainfall_for_month',
 'Rainfall_for_year',
 'Maximum_rain_per_minute',
 'Maximum_temperature',
 'Minimum_temperature',
 'Maximum_humidity',
 'Minimum_humidity',
 'Maximum_pressure',
 'Minimum_pressure',
 'Maximum_windspeed',
 'Maximum_gust_speed',
 'Maximum_heat_index',
 'Diff_pressure']

In [22]:
df[cols] = df[cols].apply(lambda x: x.astype(float))             #changing datatypes to float

In [23]:
df.info()           #successfully changed

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3903 entries, 0 to 3902
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Date                      3901 non-null   datetime64[ns]
 1   Average_temperature       3903 non-null   float64       
 2   Average_humidity          3903 non-null   float64       
 3   Average_dewpoint          3903 non-null   float64       
 4   Average_barometer         3903 non-null   float64       
 5   Average_windspeed         3903 non-null   float64       
 6   Average_gust_speed        3903 non-null   float64       
 7   Average_direction_degree  3903 non-null   float64       
 8   Rainfall_for_month        3903 non-null   float64       
 9   Rainfall_for_year         3903 non-null   float64       
 10  Maximum_rain_per_minute   3903 non-null   float64       
 11  Maximum_temperature       3903 non-null   float64       
 12  Minimum_temperature 

### To fix incorrect year in Date column :
This is most challenging task in this dataset. In our raw dataset, Date column is corrupted. We can see only 2022 year throughout the year. But the given data is day wise weather attributes from 2022 to July 2033 (predicted
data). We have to change year from 2022 to 2033.

First we will assign target_date as '2022-01-01'. Whenever we will find 2022-01-01, we will increment a year.
Replace for all rows with a new year,till we encounter '2022-01-01' and continue till the end. 

In [24]:
target_date = '2022-01-01'                         
target_date = pd.to_datetime(target_date)
index = df.index[df['Date'] == target_date].tolist()
year = 2021
max_year = 2033

for i in range(index[0], df.shape[0]):
    if df.at[i, 'Date'] == target_date:
        year += 1
        if year > max_year:
            break
    df.at[i, 'Date'] = df.at[i, 'Date'].replace(year=year)

In [25]:
df.tail(50)

Date  Average_temperature  Average_humidity  Average_dewpoint  \
3853 2033-06-10                 51.9              40.0              27.6   
3854 2033-06-11                 55.0              43.0              30.2   
3855 2033-06-12                 61.5              39.0              31.7   
3856 2033-06-13                 61.2              45.0              37.6   
3857 2033-06-14                 65.7              26.0              25.6   
3858 2033-06-15                 59.9              30.0              25.3   
3859 2033-06-16                 65.3              28.0              27.2   
3860 2033-06-17                 64.4              28.0              26.1   
3861 2033-06-18                 53.2              56.0              36.6   
3862 2033-06-19                 48.8              81.0              43.1   
3863 2033-06-20                 56.3              49.0              31.2   
3864 2033-06-21                 58.9              48.0              38.0   
3865 2033-06-22                 57.1              62.0              43.2   
3866 2033-06-23                 58.8              64.0              45.1   
3867 2033-06-24                 59.7              63.0              44.1   
3868 2033-06-25                 62.6              50.0              38.9   
3869 2033-06-26                 57.8              66.0              46.0   
3870 2033-06-27                 64.5              36.0              35.1   
3871 2033-06-28                 66.1              34.0              36.1   
3872 2033-06-29                 65.3              38.0              34.7   
3873 2033-06-30                 58.6              34.0              28.3   
3874 2033-07-01                 59.9              33.0              26.7   
3875 2033-07-02                 64.2              34.0              29.5   
3876 2033-07-03                 67.4              37.0              37.7   
3877 2033-07-04                 62.2              55.0              44.0   
3878 2033-07-05                 61.0              57.0              42.7   
3879 2033-07-06                 65.1              45.0              38.5   
3880 2033-07-07                 68.6              29.0              28.4   
3881 2033-07-08                 65.5              19.0              17.3   
3882 2033-07-09                 64.1              33.0              30.8   
3883 2033-07-10                 68.7              32.0              24.3   
3884 2033-07-11                 67.0              29.0              32.8   
3885 2033-07-12                 68.3              34.0              35.6   
3886 2033-07-13                 73.4              29.0              38.3   
3887 2033-07-14                 62.4              56.0              45.2   
3888 2033-07-15                 62.3              65.0              49.6   
3889 2033-07-16                 65.7              63.0              52.8   
3890 2033-07-17                 65.7              63.0              52.8   
3891 2033-07-18                 76.3              24.0              35.5   
3892 2033-07-19                 66.6              38.0              39.0   
3893 2033-07-20                 70.1              35.0              37.4   
3894 2033-07-21                 65.3              60.0              49.8   
3895 2033-07-22                 65.4              52.0              42.7   
3896 2033-07-23                 67.1              48.0              45.2   
3897 2033-07-24                 64.1              62.0              49.8   
3898 2033-07-25                 62.8              60.0              48.1   
3899 2033-07-26                 60.6              68.0              48.9   
3900 2033-07-27                 61.7              64.0              47.4   
3901 2033-07-28                 60.5              61.0              45.3   
3902 2033-07-28                 60.5              61.0              45.3   

      Average_barometer  Average_windspeed  Average_gust_speed  \
3853               30.0                6.7                10.5   
38

In [26]:
random_date = '2024-02-29'
rows = df.loc[df['Date'] == random_date]                        #to check for any uncommon values present 
rows

Empty DataFrame
Columns: [Date, Average_temperature, Average_humidity, Average_dewpoint, Average_barometer, Average_windspeed, Average_gust_speed, Average_direction_degree, Rainfall_for_month, Rainfall_for_year, Maximum_rain_per_minute, Maximum_temperature, Minimum_temperature, Maximum_humidity, Minimum_humidity, Maximum_pressure, Minimum_pressure, Maximum_windspeed, Maximum_gust_speed, Maximum_heat_index, Diff_pressure]
Index: []

In [27]:
df['Date']           #exploring date column

0      2022-01-01
1      2022-01-02
2      2022-01-03
3      2022-01-04
4      2022-01-05
          ...    
3898   2033-07-25
3899   2033-07-26
3900   2033-07-27
3901   2033-07-28
3902   2033-07-28
Name: Date, Length: 3903, dtype: datetime64[ns]

### Handling Duplicates :

In [28]:
df[df.duplicated()] #to show duplicate records

Date  Average_temperature  Average_humidity  Average_dewpoint  \
3902 2033-07-28                 60.5              61.0              45.3   

      Average_barometer  Average_windspeed  Average_gust_speed  \
3902               29.7                4.0                 6.2   

      Average_direction_degree  Rainfall_for_month  Rainfall_for_year  \
3902                     248.0                0.35               4.49   

      Maximum_rain_per_minute  Maximum_temperature  Minimum_temperature  \
3902                      0.0                 75.6                 46.0   

      Maximum_humidity  Minimum_humidity  Maximum_pressure  Minimum_pressure  \
3902              94.0              35.0            29.792            29.675   

      Maximum_windspeed  Maximum_gust_speed  Maximum_heat_index  Diff_pressure  
3902               17.3                26.5                77.6          0.117

In [29]:
#dropping all duplicates from df
df.drop_duplicates(inplace=True)
df.tail()

Date  Average_temperature  Average_humidity  Average_dewpoint  \
3897 2033-07-24                 64.1              62.0              49.8   
3898 2033-07-25                 62.8              60.0              48.1   
3899 2033-07-26                 60.6              68.0              48.9   
3900 2033-07-27                 61.7              64.0              47.4   
3901 2033-07-28                 60.5              61.0              45.3   

      Average_barometer  Average_windspeed  Average_gust_speed  \
3897               29.6                3.6                 5.8   
3898               29.7                2.5                 4.0   
3899               29.8                1.7                 2.9   
3900               29.9                2.2                 4.0   
3901               29.7                4.0                 6.2   

      Average_direction_degree  Rainfall_for_month  Rainfall_for_year  \
3897                     240.0                0.24               4.38   
3898                     242.0                0.33               4.47   
3899                     357.0                0.33               4.47   
3900                      66.0                0.33               4.47   
3901                     248.0                0.35               4.49   

      Maximum_rain_per_minute  Maximum_temperature  Minimum_temperature  \
3897                      0.0                 74.9                 55.3   
3898                      0.0                 69.2                 55.1   
3899                      0.0                 71.9                 50.5   
3900                      0.0                 77.3                 43.6   
3901                      0.0                 75.6                 46.0   

      Maximum_humidity  Minimum_humidity  Maximum_pressure  Minimum_pressure  \
3897              86.0              35.0            29.686            29.577   
3898              90.0              36.0            29.781            29.645   
3899              90.0              40.0            29.930            29.745   
3900              96.0              35.0            29.941            29.781   
3901              94.0              35.0            29.792            29.675   

      Maximum_windspeed  Maximum_gust_speed  Maximum_heat_index  Diff_pressure  
3897               15.0                25.3                77.4          0.109  
3898                8.1                17.3                77.5          0.136  
3899               11.5                15.0                77.5          0.185  
3900               13.8                18.4                78.2          0.160  
3901               17.3                26.5                77.6          0.117

Here we have ssuccessfully dropped the duplicates!

Now we will check for the leap year data 

In [30]:
df[(df['Date'].dt.month == 2) & (df['Date'].dt.day == 29)]

Empty DataFrame
Columns: [Date, Average_temperature, Average_humidity, Average_dewpoint, Average_barometer, Average_windspeed, Average_gust_speed, Average_direction_degree, Rainfall_for_month, Rainfall_for_year, Maximum_rain_per_minute, Maximum_temperature, Minimum_temperature, Maximum_humidity, Minimum_humidity, Maximum_pressure, Minimum_pressure, Maximum_windspeed, Maximum_gust_speed, Maximum_heat_index, Diff_pressure]
Index: []

We will change format of date column to date-only.

In [31]:
df['Date'].dtype
df['Date'] = df['Date'].dt.date #to convert the datetime column to a date-only format

### Dealing with null values :

In [32]:
df.isnull().sum()            #to check for null values

Date                        2
Average_temperature         0
Average_humidity            0
Average_dewpoint            0
Average_barometer           0
Average_windspeed           0
Average_gust_speed          0
Average_direction_degree    0
Rainfall_for_month          0
Rainfall_for_year           0
Maximum_rain_per_minute     0
Maximum_temperature         0
Minimum_temperature         0
Maximum_humidity            0
Minimum_humidity            0
Maximum_pressure            0
Minimum_pressure            0
Maximum_windspeed           0
Maximum_gust_speed          0
Maximum_heat_index          0
Diff_pressure               0
dtype: int64

In [33]:
df[df.isna().any(axis=1)]            #exploring for null values

Date  Average_temperature  Average_humidity  Average_dewpoint  \
2366  NaT                 38.1              36.0              12.3   
3751  NaT                 37.6              31.0               8.7   

      Average_barometer  Average_windspeed  Average_gust_speed  \
2366               29.8               11.5                17.0   
3751               29.8                8.0                11.6   

      Average_direction_degree  Rainfall_for_month  Rainfall_for_year  \
2366                      77.0                0.16               0.36   
3751                     236.0                0.48               0.49   

      Maximum_rain_per_minute  Maximum_temperature  Minimum_temperature  \
2366                      0.0                 45.9                 27.9   
3751                      0.0                 48.1                 26.1   

      Maximum_humidity  Minimum_humidity  Maximum_pressure  Minimum_pressure  \
2366              73.0              22.0            30.023            29.702   
3751              57.0              20.0            29.972            29.647   

      Maximum_windspeed  Maximum_gust_speed  Maximum_heat_index  Diff_pressure  
2366               29.9                39.1                45.9          0.321  
3751               19.6                31.1                48.1          0.325

In [34]:
df.iloc[[2365,2366,2367,3750,3751,3752]]     #exploring more about null values

Date  Average_temperature  Average_humidity  Average_dewpoint  \
2365  2029-02-28                 42.2              25.0               7.1   
2366         NaT                 38.1              36.0              12.3   
2367  2029-03-01                 36.4              30.0               7.6   
3750  2033-02-28                 37.7              30.0               8.3   
3751         NaT                 37.6              31.0               8.7   
3752  2033-03-01                 28.7              66.0              17.8   

      Average_barometer  Average_windspeed  Average_gust_speed  \
2365               29.9                8.7                13.2   
2366               29.8               11.5                17.0   
2367               30.0                7.9                12.5   
3750               30.1               11.0                15.9   
3751               29.8                8.0                11.6   
3752               29.7                4.1                 6.5   

      Average_direction_degree  Rainfall_for_month  Rainfall_for_year  \
2365                      58.0                0.16               0.36   
2366                      77.0                0.16               0.36   
2367                      23.0                0.00               0.21   
3750                     224.0                0.48               0.49   
3751                     236.0                0.48               0.49   
3752                      34.0                0.00               0.49   

      Maximum_rain_per_minute  Maximum_temperature  Minimum_temperature  \
2365                      0.0                 53.6                 31.5   
2366                      0.0                 45.9                 27.9   
2367                      0.0                 47.0                 24.5   
3750                      0.0                 47.0                 29.5   
3751                      0.0                 48.1                 26.1   
3752                      0.0                 38.2                 17.1   

      Maximum_humidity  Minimum_humidity  Maximum_pressure  Minimum_pressure  \
2365              44.0              11.0            29.942            29.749   
2366              73.0              22.0            30.023            29.702   
2367              51.0              22.0            30.125            29.817   
3750              38.0              21.0            30.194            29.970   
3751              57.0              20.0            29.972            29.647   
3752              90.0              36.0            29.903            29.622   

      Maximum_windspeed  Maximum_gust_speed  Maximum_heat_index  Diff_pressure  
2365               25.3                36.8                53.6          0.193  
2366               29.9                39.1                45.9          0.321  
2367               24.2                34.5                47.0          0.308  
3750               24.2                42.6                47.0          0.224  
3751               19.6                31.1                48.1          0.325  
3752               16.1                23.0                38.2          0.281

As we can see above, null values are for the date where 29th day was present in the feb for non leap year. We can drop it.

In [35]:
df.dropna(inplace=True)    #dropping rows having NaT values

In [36]:
df[df.isna().any(axis=1)]             #successfully dropped

Empty DataFrame
Columns: [Date, Average_temperature, Average_humidity, Average_dewpoint, Average_barometer, Average_windspeed, Average_gust_speed, Average_direction_degree, Rainfall_for_month, Rainfall_for_year, Maximum_rain_per_minute, Maximum_temperature, Minimum_temperature, Maximum_humidity, Minimum_humidity, Maximum_pressure, Minimum_pressure, Maximum_windspeed, Maximum_gust_speed, Maximum_heat_index, Diff_pressure]
Index: []

We have completed all preprocessing and data cleaning. Final shape of our dataset is: 

In [37]:
df.shape

(3900, 21)

## To Export Clean Dataset into csv File :

In [38]:
df.to_csv('weather_datasetcleaned.csv', index=False, encoding='utf-8')

***
## Conclusion :

- We have done all pre-processing using python alone with the help of Pandas and Numpy. 
- Corruped data, duplicates and missed data handled and fixed successfully.